colexec: add DiskQueueCacheModes to the DiskQueue

asubiotto opened the pull request:
Refer to the commits for details. In a nutshell, this change allows us to reduce the partition cache size without sacrificing performance for more flexibility when partitioning external sorts and hash joins. Additionally, this commit adds the functionality to close write partition caches when the hash joiner calls `PartitionDiskQueue.CloseAllOpenWriteFileDescriptors`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45501)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
Nice! Got some nits and one question.

> Reviewed 2 of 2 files at r1, 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45501#-:-M182H5m5uEF122pRi2g:bfpg76e)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colcontainer/diskqueue.go, line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45501#-M18-1Kw7428y8ZkdR0t:-M18-1Kw7428y8ZkdR0u:bvu7zdu) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c80808ac5e20b3a23289b818fcede267968331d/pkg/sql/colcontainer/diskqueue.go#L129)):*
> ```Go
> }
> 
> // diskQueueState is a simple state that records the current state of the disk
> ```

nit: too many "states" in the first line. Maybe simply "diskQueueState describes the current state of the disk..."
___
*[pkg/sql/colcontainer/diskqueue.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45501#-M18-Eh13tdVGvw4xxxU:-M18-Eh13tdVGvw4xxxV:bb68hpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c80808ac5e20b3a23289b818fcede267968331d/pkg/sql/colcontainer/diskqueue.go#L150)):*
> ```Go
> // A diskQueue will never use more memory than cfg.BufferSizeBytes, but not all
> // the available memory will be used to buffer only writes. Refer to the
> // DiskQueuCacheMode comment as to how cfg.BufferSizeBytes is divided in each
> ```

nit: `s/Queu/Queue/g`.
___
*[pkg/sql/colcontainer/diskqueue.go, line 215 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45501#-M18-gZx5fsU7m1BuM63:-M18-gZx5fsU7m1BuM64:b4khqy2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c80808ac5e20b3a23289b818fcede267968331d/pkg/sql/colcontainer/diskqueue.go#L215)):*
> ```Go
> 	// DefaultBufferSizeBytesDefaultCacheMode is the default buffer size used when
> 	// the DiskQueue is in DiskQueueCacheModeDefault.
> 	// These values were chosen by running BenchmarkQueue.
> ```

nit: this line refers to the first and the third values, probably we should change the order or explicitly call out what "these" refers to.
___
*[pkg/sql/colcontainer/diskqueue.go, line 247 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45501#-M18-y2zEShGeLzeswHB:-M18-y2zEShGeLzeswHC:b-y4p2r7) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c80808ac5e20b3a23289b818fcede267968331d/pkg/sql/colcontainer/diskqueue.go#L247)):*
> ```Go
> 	// DiskQueueCacheModeClearAndReuseCache is the same as
> 	// DiskQueueCacheModeReuseCache with the additional behavior that when a
> 	// coldata.ZeroBatch is Enqueued, the cache will be release to the GC.
> ```

nit: `s/release/released/g`.
___
*[pkg/sql/colcontainer/diskqueue.go, line 318 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45501#-M180aekDVa4a2n4kHeV:-M180aekDVa4a2n4kHeW:b-c53vn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c80808ac5e20b3a23289b818fcede267968331d/pkg/sql/colcontainer/diskqueue.go#L318)):*
> ```Go
> 	// Refer to the DiskQueueCacheMode comment for why this division of
> 	// BufferSizeBytes.
> 	if d.cfg.CacheMode == DiskQueueCacheModeDefault {
> ```

nit: I'd expect for the "default" mode to be present in the instantiation of the disk queue above, and here we would override for non-default modes.
___
*[pkg/sql/colcontainer/diskqueue.go, line 469 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45501#-M181-qF0364TNe-DMWZ:-M181-qF0364TNe-DMW_:b-2via41) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c80808ac5e20b3a23289b818fcede267968331d/pkg/sql/colcontainer/diskqueue.go#L469)):*
> ```Go
> 	d.numBufferedBatches++
> 
> 	// Only buffer up to a third of the available buffer space, we need another
> ```

nit: this comment needs an adjustment (maybe it can be removed entirely).
___
*[pkg/sql/colcontainer/diskqueue_test.go, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45501#-M181B_KAB-4FrvlKO26:-M181B_KAB-4FrvlKO27:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c80808ac5e20b3a23289b818fcede267968331d/pkg/sql/colcontainer/diskqueue_test.go#L49)):*
> ```Go
> 			alwaysCompress := rng.Float64() < 0.5
> 			diskQueueCacheMode := colcontainer.DiskQueueCacheModeDefault
> 			// testReuseCache will test the reuse cache modes
> ```

nit: missing period.
___
*[pkg/sql/colcontainer/partitionedqueue_test.go, line 168 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45501#-M181d1T7KyJB_91ix_6:-M181d1T7KyJB_91ix_7:b-gt0rn1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a986987ca7d79ff8383a59da91ec568d93d7f29/pkg/sql/colcontainer/partitionedqueue_test.go#L168)):*
> ```Go
> 	t.Run(fmt.Sprintf("Sort/maxPartitions=%d/numRepartitions=%d", maxPartitions, numRepartitions), func(t *testing.T) {
> 		queueCfg.CacheMode = colcontainer.DiskQueueCacheModeReuseCache
> 		queueCfg.BufferSizeBytes = colcontainer.DefaultBufferSizeBytesReuseCacheMode
> ```

Can we leave this up to `EnsureDefaults`?
___
*[pkg/sql/colexec/routers.go, line 399 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45501#-M1826_Y0CyWAgi1QTvM:-M1826_ZCb1A8DhZqMFQ:b-bxnstd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a986987ca7d79ff8383a59da91ec568d93d7f29/pkg/sql/colexec/routers.go#L399)):*
> ```Go
> ) (*HashRouter, []Operator) {
> 	if diskQueueCfg.CacheMode != colcontainer.DiskQueueCacheModeDefault {
> 		execerror.VectorizedInternalPanic(errors.Errorf("hash router instantiated incompatible disk queue cache mode: %d", diskQueueCfg.CacheMode))
> ```

nit: `s/instantiated/instantiated with/g`.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
Clicked on `Resolve` instead of commenting `Done`, let's see if that's useful?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45501#-:-M1CX0Xq0-VLC60WkhPW:bbn36sm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colcontainer/partitionedqueue_test.go, line 168 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45501#-M181d1T7KyJB_91ix_6:-M1CRblp3AvegjeXFTiq:b-ld5lmi) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a986987ca7d79ff8383a59da91ec568d93d7f29/pkg/sql/colcontainer/partitionedqueue_test.go#L168)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Can we leave this up to `EnsureDefaults`?
</blockquote></details>

We could, but we'd have to set it to 0 before calling `EnsureDefaults` for the default buffer size for the given cache mode to be used. Maybe we could add a `SetDefaultBufferSizeForCacheMode`. Actually I like that. Done.


<!-- Sent from Reviewable.io -->

Commit:
colexec: use DiskQueueCacheModes to optimize disk operators

The external sorter and hash joiner may use the two ReuseCache modes
to reduce the cache size without sacrificing performance given their access
patterns per partition of write everything then read everything.

The external hash joiner additionally uses the clear cache mode in order to
automatically closed unused partition caches while it is repartitioning.

Release note: None (external hash joins and sorts are not enabled by default
yet)

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Resolving nits is helpful - it removes the clutter :)

> Reviewed 2 of 7 files at r3, 5 of 5 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45501#-:-M1Dmwz_Cd0MDZ-BnoAo:b-1268ck)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

Commit:
colcontainer: add DiskQueueCacheModes to the DiskQueue

This allows the caller to specify whether they want to reuse the DiskQueue's
cache. This comes with the limitation that all writes must happen before all
reads, which is the case for the ExternalHashJoiner and ExternalSorter (but
not the HashRouter). This allows us to use less memory per partition. An
additional mode has also been added that releases the cache to the GC. This
is useful in the case of an ExternalHashJoinerRepartition, where if a
repartition happens, the other partitions at that level are untouched until
the repartitioned partition has been repartitioned n times until it is fully
joined and emitted.

Release note: None (internal code change with no observable external changes)

asubiotto commented:
bors r=yuzefovich

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45501)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colexec/external_sort.go, line 174 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45501#-M1Ls0or1vOc-dOJFW3Y:-M1Ls0or1vOc-dOJFW3Z:boy366o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed32a68b56692a2c04fb2eca464113890c2b81b2/pkg/sql/colexec/external_sort.go#L174)):*
> ```Go
> 		// that mergers will use).
> 		batchMemSize := estimateBatchSizeBytes(inputTypes, int(coldata.BatchSize()))
> 		maxNumberPartitions = (int(memoryLimit) - batchMemSize) / (diskQueueCfg.BufferSizeBytes + batchMemSize)
> ```

We need to be a little more careful when rebasing.


<!-- Sent from Reviewable.io -->
